File: software/examples/vbnet/ExampleCallback.vb
Imports System
Imports Tinkerforge

Module ExampleCallback
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change to your UID

    ' Callback subroutine for motion detected callback
    Sub MotionDetectedCB(ByVal sender As BrickletMotionDetector)
        Console.WriteLine("Motion Detected")
    End Sub

    ' Callback subroutine for detection cycle ended callback
    Sub DetectionCycleEndedCB(ByVal sender As BrickletMotionDetector)
        Console.WriteLine("Detection Cycle Ended (next detection possible in ~3 seconds)")
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim md As New BrickletMotionDetector(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Register motion detected callback to subroutine MotionDetectedCB
        AddHandler md.MotionDetected, AddressOf MotionDetectedCB

        ' Register detection cycle ended callback to subroutine DetectionCycleEndedCB
        AddHandler md.DetectionCycleEnded, AddressOf DetectionCycleEndedCB

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module

File: software/examples/delphi/ExampleCallback.pas
program ExampleCallback;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletMotionDetector;

type
  TExample = class
  private
    ipcon: TIPConnection;
    md: TBrickletMotionDetector;
  public
    procedure MotionDetectedCB(sender: TBrickletMotionDetector);
    procedure DetectionCycleEndedCB(sender: TBrickletMotionDetector);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'XYZ'; { Change to your UID }

var
  e: TExample;

{ Callback function for detected motion }
procedure TExample.MotionDetectedCB(sender: TBrickletMotionDetector);
begin
  WriteLn('Motion Detected');
end;

{ Callback function for end of detection cycle }
procedure TExample.DetectionCycleEndedCB(sender: TBrickletMotionDetector);
begin
  WriteLn('Detection Cycle Ended (next detection possible in ~3 seconds)');
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  md := TBrickletMotionDetector.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Register detected callback to function MotionDetectedCB }
  md.OnMotionDetected := {$ifdef FPC}@{$endif}MotionDetectedCB;

  { Register detection cycle ended callback to function DetectionCycleEndedCB }
  md.OnDetectionCycleEnded := {$ifdef FPC}@{$endif}DetectionCycleEndedCB;

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.
